reasoner/planner/ReasonerPlanner.java

```diff
@@ -0,0 +1,144 @@
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+package com.vaticle.typedb.core.reasoner.planner;
+
+import com.vaticle.typedb.common.collection.Pair;
+import com.vaticle.typedb.core.concept.ConceptManager;
+import com.vaticle.typedb.core.logic.LogicManager;
+import com.vaticle.typedb.core.logic.resolvable.Concludable;
+import com.vaticle.typedb.core.logic.resolvable.Resolvable;
+import com.vaticle.typedb.core.logic.resolvable.ResolvableConjunction;
+import com.vaticle.typedb.core.logic.resolvable.Retrievable;
+import com.vaticle.typedb.core.pattern.variable.ThingVariable;
+import com.vaticle.typedb.core.traversal.TraversalEngine;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public abstract class ReasonerPlanner {
+    protected final Map<Pair<ResolvableConjunction, Set<Identifier.Variable.Retrievable>>, Plan<Resolvable<?>>> planCache;
+    protected final ConceptManager conceptMgr;
+    protected final TraversalEngine traversalEng;
+    protected final LogicManager logicMgr;
+    private final HashMap<ResolvableConjunction, Set<Resolvable<?>>> compiled;
+
+    public ReasonerPlanner(TraversalEngine traversalEng, ConceptManager conceptMgr, LogicManager logicMgr) {
+        planCache = new HashMap<>();
+        this.traversalEng = traversalEng;
+        this.conceptMgr = conceptMgr;
+        this.logicMgr = logicMgr;
+        this.compiled = new HashMap<>();
+    }
+
+    public static ReasonerPlanner create(TraversalEngine traversalEng, ConceptManager conceptMgr, LogicManager logicMgr) {
+        return new GreedyCostSearch.OldPlannerEmulator(traversalEng, conceptMgr, logicMgr);
+    }
+
+    public Set<Resolvable<?>> compile(ResolvableConjunction conjunction) {
+        if (!compiled.containsKey(conjunction)) {
+            synchronized (compiled) {
+                if (!compiled.containsKey(conjunction)){
+                    Set<Concludable> concludablesTriggeringRules = iterate(conjunction.positiveConcludables())
+                            .filter(concludable -> !logicMgr.applicableRules(concludable).isEmpty())
+                            .toSet();
+                    Set<Resolvable<?>> resolvables = new HashSet<>();
+                    resolvables.addAll(concludablesTriggeringRules);
+                    resolvables.addAll(Retrievable.extractFrom(conjunction.pattern(), concludablesTriggeringRules));
+                    resolvables.addAll(conjunction.negations());
+                    compiled.put(conjunction, resolvables);
+                }
+            }
+        }
+        return compiled.get(conjunction);
+    }
+
+    public Plan<Resolvable<?>> getPlan(ResolvableConjunction conjunction, Set<Identifier.Variable.Retrievable> bounds) {
```

Comment:
We only really use `get` for real getters, and even then only sometimes (to match APIs that are available to the user on the client-side concept API). On the server side this would just be `plan(...)`